File: app/api/aica_api/services/behavior/driver_model.py
"""Driver behavior model (T009) — additive per-tick driver state progression.

Pure, deterministic, side-effect-free.  Same (profile, state, context) →
same DriverUpdate every call.

Rate model (R1):
  drowsiness += (base_growth + night_add * is_night
                 + monotony_add * is_monotonous
                 + jam_add * is_traffic_jam) * tick_seconds / 60
  fatigue    += (base_growth + continuous_add * (continuous_min >= 60)
                 + mountain_add * is_mountain_road
                 + jam_add * is_traffic_jam) * tick_seconds / 60
  attention  += (base_recovery - monotony_drop * is_monotonous
                 - drowsiness_drop_factor * drowsiness / 100
                 + active_content_recovery * active_content) * tick_seconds / 60

All levels clamped [0, 100].  Component deltas recorded in DriverDelta.
"""

from __future__ import annotations

from dataclasses import dataclass

from aica_api.models.profile import DriverModelProfile


@dataclass
class DriverState:
    """Instantaneous numeric driver state (all in [0, 100] range)."""

    drowsiness: float
    fatigue: float
    attention: float


@dataclass
class DriverDelta:
    """Component-level deltas for this tick (for the evidence trace)."""

    drowsiness_base: float
    drowsiness_night: float
    drowsiness_monotony: float
    drowsiness_jam: float

    fatigue_base: float
    fatigue_continuous: float
    fatigue_mountain: float
    fatigue_jam: float

    attention_base_recovery: float
    attention_monotony_drop: float
    attention_drowsiness_drop: float
    attention_active_content: float


@dataclass
class DriverUpdate:
    """Result of one tick: previous state, deltas, next state."""

    previous: DriverState
    delta: DriverDelta
    next: DriverState


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def advance_driver_state(
    profile: DriverModelProfile,
    current: DriverState,
    tick_seconds: int,
    *,
    is_night: bool,
    is_monotonous: bool,
    is_traffic_jam: bool,
    is_mountain_road: bool,
    continuous_driving_min: float,
    active_content: bool = False,
) -> DriverUpdate:
    """Advance driver state by one tick.

    Args:
        profile:               Validated DriverModelProfile.
        current:               State at the START of this tick.
        tick_seconds:          Duration of this tick in seconds.
        is_night:              True if driving at night.
        is_monotonous:         True if segment is monotonous (highway / normal_road).
        is_traffic_jam:        True if a traffic jam is active.
        is_mountain_road:      True if segment is mountain_road.
        continuous_driving_min: Elapsed continuous driving minutes (for >60-min term).
        active_content:        True if a recovery-content action is active.

    Returns:
        A DriverUpdate with previous, delta, and next states.
    """
    scale = tick_seconds / 60.0  # per-minute rates → per-tick

    dm = profile.drowsiness_model
    fm = profile.fatigue_model
    am = profile.attention_model

    # ── Drowsiness deltas ─────────────────────────────────────────────────
    d_base = dm.base_growth_per_min * scale
    d_night = dm.night_add_per_min * scale if is_night else 0.0
    d_monotony = dm.monotony_add_per_min * scale if is_monotonous else 0.0
    d_jam = dm.traffic_jam_add_per_min * scale if is_traffic_jam else 0.0

    new_drowsiness = _clamp(current.drowsiness + d_base + d_night + d_monotony + d_jam)

    # ── Fatigue deltas ────────────────────────────────────────────────────
    f_base = fm.base_growth_per_min * scale
    f_continuous = (
        fm.continuous_driving_add_per_min_after_60_min * scale
        if continuous_driving_min >= 60.0
        else 0.0
    )
    f_mountain = fm.mountain_road_add_per_min * scale if is_mountain_road else 0.0
    f_jam = fm.traffic_jam_add_per_min * scale if is_traffic_jam else 0.0

    new_fatigue = _clamp(current.fatigue + f_base + f_continuous + f_mountain + f_jam)

    # ── Attention deltas ──────────────────────────────────────────────────
    a_recovery = am.base_recovery_per_min * scale
    a_monotony_drop = am.monotony_drop_per_min * scale if is_monotonous else 0.0
    # drowsiness_drop_factor is applied to normalized drowsiness (0→1)
    a_drowsiness_drop = am.drowsiness_drop_factor * (current.drowsiness / 100.0) * scale
    a_content = am.active_content_recovery_per_min * scale if active_content else 0.0

    new_attention = _clamp(
        current.attention + a_recovery - a_monotony_drop - a_drowsiness_drop + a_content
    )

    delta = DriverDelta(
        drowsiness_base=d_base,
        drowsiness_night=d_night,
        drowsiness_monotony=d_monotony,
        drowsiness_jam=d_jam,
        fatigue_base=f_base,
        fatigue_continuous=f_continuous,
        fatigue_mountain=f_mountain,
        fatigue_jam=f_jam,
        attention_base_recovery=a_recovery,
        attention_monotony_drop=-a_monotony_drop,
        attention_drowsiness_drop=-a_drowsiness_drop,
        attention_active_content=a_content,
    )

    return DriverUpdate(
        previous=current,
        delta=delta,
        next=DriverState(
            drowsiness=new_drowsiness,
            fatigue=new_fatigue,
            attention=new_attention,
        ),
    )


def apply_rest_recovery(
    profile: DriverModelProfile,
    current: DriverState,
    rest_type: str,
) -> DriverState:
    """Apply rest recovery to the current driver state.

    Args:
        profile:   DriverModelProfile (provides recovery amounts).
        current:   State before rest.
        rest_type: "short" or "long".

    Returns:
        Recovered DriverState (drowsiness/fatigue reduced, attention improved).
    """
    rm = profile.recovery_model
    if rest_type == "long":
        d_rec = rm.long_rest_drowsiness_recovery
        f_rec = rm.long_rest_fatigue_recovery
    else:  # "short" (default)
        d_rec = rm.short_rest_drowsiness_recovery
        f_rec = rm.short_rest_fatigue_recovery

    return DriverState(
        drowsiness=_clamp(current.drowsiness - d_rec),
        fatigue=_clamp(current.fatigue - f_rec),
        # Attention improves proportionally to drowsiness recovery
        attention=_clamp(current.attention + d_rec * 0.5),
    )


# ---------------------------------------------------------------------------
# Private helpers
# ---------------------------------------------------------------------------


def _clamp(value: float, lo: float = 0.0, hi: float = 100.0) -> float:
    """Clamp value to [lo, hi]."""
    return max(lo, min(hi, value))

File: htmlapp/frontend/src/engine/behavior/driver_model.ts
/**
 * Driver behavior model — additive per-tick driver state progression.
 *
 * Ported from `app/api/aica_api/services/behavior/driver_model.py`
 * (behavior-of-record). Pure, deterministic, side-effect-free: same
 * (profile, state, tick args) → same DriverUpdate every call.
 *
 * Rate model (R1):
 *   drowsiness += (base_growth + night_add * is_night
 *                  + monotony_add * is_monotonous
 *                  + jam_add * is_traffic_jam) * tick_seconds / 60
 *   fatigue    += (base_growth + continuous_add * (continuous_min >= 60)
 *                  + mountain_add * is_mountain_road
 *                  + jam_add * is_traffic_jam) * tick_seconds / 60
 *   attention  += (base_recovery - monotony_drop * is_monotonous
 *                  - drowsiness_drop_factor * drowsiness / 100
 *                  + active_content_recovery * active_content) * tick_seconds / 60
 *
 * All levels clamped [0, 100]. Component deltas recorded in DriverDelta.
 *
 * Only the exported function identifiers (advanceDriverState, applyRestRecovery)
 * are camelCased. All object keys cross the parity boundary and are preserved
 * byte-for-byte from the Python (snake_case).
 *
 * The Python's keyword-only tick args (`is_night`, `is_monotonous`, ...) become
 * a trailing options object here, mirroring the same convention used for
 * recovery's `at_rest_spot` → `{ atRestSpot }`.
 */

// ---------------------------------------------------------------------------
// Public data types
// ---------------------------------------------------------------------------

/** Instantaneous numeric driver state (all in [0, 100] range). */
export type DriverState = {
  drowsiness: number
  fatigue: number
  attention: number
}

/** Component-level deltas for this tick (for the evidence trace). */
export type DriverDelta = {
  drowsiness_base: number
  drowsiness_night: number
  drowsiness_monotony: number
  drowsiness_jam: number

  fatigue_base: number
  fatigue_continuous: number
  fatigue_mountain: number
  fatigue_jam: number

  attention_base_recovery: number
  attention_monotony_drop: number
  attention_drowsiness_drop: number
  attention_active_content: number
}

/** Result of one tick: previous state, deltas, next state. */
export type DriverUpdate = {
  previous: DriverState
  delta: DriverDelta
  next: DriverState
}

// ---------------------------------------------------------------------------
// Profile types (mirror aica_api.models.profile; not present in api/types.ts)
// ---------------------------------------------------------------------------

export type DrowsinessModel = {
  base_growth_per_min: number
  night_add_per_min: number
  monotony_add_per_min: number
  traffic_jam_add_per_min: number
}

export type FatigueModel = {
  base_growth_per_min: number
  continuous_driving_add_per_min_after_60_min: number
  mountain_road_add_per_min: number
  traffic_jam_add_per_min: number
}

export type AttentionModel = {
  base_recovery_per_min: number
  monotony_drop_per_min: number
  drowsiness_drop_factor: number
  active_content_recovery_per_min: number
}

export type RecoveryModel = {
  short_rest_drowsiness_recovery: number
  short_rest_fatigue_recovery: number
  long_rest_drowsiness_recovery: number
  long_rest_fatigue_recovery: number
}

export type DriverModelProfile = {
  id: string
  drowsiness_model: DrowsinessModel
  fatigue_model: FatigueModel
  attention_model: AttentionModel
  recovery_model: RecoveryModel
}

/** Keyword-only tick args from the Python signature, as a trailing options object. */
export type AdvanceDriverStateOptions = {
  isNight: boolean
  isMonotonous: boolean
  isTrafficJam: boolean
  isMountainRoad: boolean
  continuousDrivingMin: number
  activeContent?: boolean
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/** Advance driver state by one tick. */
export function advanceDriverState(
  profile: DriverModelProfile,
  current: DriverState,
  tickSeconds: number,
  options: AdvanceDriverStateOptions,
): DriverUpdate {
  const {
    isNight,
    isMonotonous,
    isTrafficJam,
    isMountainRoad,
    continuousDrivingMin,
    activeContent = false,
  } = options

  const scale = tickSeconds / 60.0 // per-minute rates → per-tick

  const dm = profile.drowsiness_model
  const fm = profile.fatigue_model
  const am = profile.attention_model

  // ── Drowsiness deltas ─────────────────────────────────────────────────
  const dBase = dm.base_growth_per_min * scale
  const dNight = isNight ? dm.night_add_per_min * scale : 0.0
  const dMonotony = isMonotonous ? dm.monotony_add_per_min * scale : 0.0
  const dJam = isTrafficJam ? dm.traffic_jam_add_per_min * scale : 0.0

  const newDrowsiness = clamp(current.drowsiness + dBase + dNight + dMonotony + dJam)

  // ── Fatigue deltas ────────────────────────────────────────────────────
  const fBase = fm.base_growth_per_min * scale
  const fContinuous = continuousDrivingMin >= 60.0
    ? fm.continuous_driving_add_per_min_after_60_min * scale
    : 0.0
  const fMountain = isMountainRoad ? fm.mountain_road_add_per_min * scale : 0.0
  const fJam = isTrafficJam ? fm.traffic_jam_add_per_min * scale : 0.0

  const newFatigue = clamp(current.fatigue + fBase + fContinuous + fMountain + fJam)

  // ── Attention deltas ──────────────────────────────────────────────────
  const aRecovery = am.base_recovery_per_min * scale
  const aMonotonyDrop = isMonotonous ? am.monotony_drop_per_min * scale : 0.0
  // drowsiness_drop_factor is applied to normalized drowsiness (0→1)
  const aDrowsinessDrop = am.drowsiness_drop_factor * (current.drowsiness / 100.0) * scale
  const aContent = activeContent ? am.active_content_recovery_per_min * scale : 0.0

  const newAttention = clamp(
    current.attention + aRecovery - aMonotonyDrop - aDrowsinessDrop + aContent,
  )

  const delta: DriverDelta = {
    drowsiness_base: dBase,
    drowsiness_night: dNight,
    drowsiness_monotony: dMonotony,
    drowsiness_jam: dJam,
    fatigue_base: fBase,
    fatigue_continuous: fContinuous,
    fatigue_mountain: fMountain,
    fatigue_jam: fJam,
    attention_base_recovery: aRecovery,
    attention_monotony_drop: -aMonotonyDrop,
    attention_drowsiness_drop: -aDrowsinessDrop,
    attention_active_content: aContent,
  }

  return {
    previous: current,
    delta,
    next: {
      drowsiness: newDrowsiness,
      fatigue: newFatigue,
      attention: newAttention,
    },
  }
}

/** Apply rest recovery to the current driver state. `restType`: "short" or "long". */
export function applyRestRecovery(
  profile: DriverModelProfile,
  current: DriverState,
  restType: string,
): DriverState {
  const rm = profile.recovery_model
  let dRec: number
  let fRec: number
  if (restType === 'long') {
    dRec = rm.long_rest_drowsiness_recovery
    fRec = rm.long_rest_fatigue_recovery
  } else {
    // "short" (default)
    dRec = rm.short_rest_drowsiness_recovery
    fRec = rm.short_rest_fatigue_recovery
  }

  return {
    drowsiness: clamp(current.drowsiness - dRec),
    fatigue: clamp(current.fatigue - fRec),
    // Attention improves proportionally to drowsiness recovery
    attention: clamp(current.attention + dRec * 0.5),
  }
}

// ---------------------------------------------------------------------------
// Private helpers
// ---------------------------------------------------------------------------

/** Clamp value to [lo, hi]. */
function clamp(value: number, lo = 0.0, hi = 100.0): number {
  return Math.max(lo, Math.min(hi, value))
}
